lib/memory.cpp

```diff
@@ -234,6 +236,13 @@ void Memory::store(Word addr, uint8_t data_type, Word data) {
 }
 
 void Memory::update_peripherals() {
+    //Only actually update/check peripherals every 65535 times this function is called
+    //This is since, although non-blocking, the read system call in the UART is REALLY REALLY REALLY slow
+    //FIXME does this introduce software problems? I don't think so, no chance some types that fast anyways and
```

Comment:
The comments discuss (not too well) two separate issues here.

The first is the prohibitively high cost of calling `read` on each tick. Even if there are no characters to receive, issuing a system call to do a non-blocking read has a very, very high cost that was resulting in around a 25% slowdown (from the original speed, so it would probably be closer to halving the current speed after the optimizations I made elsewhere). So we only actually call it every certain number of ticks.

The second is the problem that this optimization potentially introduces, and you're exactly bang-on about what I'm trying to say here. What if we receive two characters between updates? I don't think it matters because the kernel will be buffering those characters until we get around to calling `read`. It's not like they'd be lost if we don't "pick them up in time," or at least that's what I was thinking. Admittedly this feels fairly obvious so I likely didn't need to include that comment. :)

Let me know if there's any more confusion / if I should reword these comments / if you think that there may be an issue with my reasoning.

Reply:
Ah that makes sense, thanks